improve MDBValue type handling

tanmaykm opened the pull request:
- prevent possible invalid ptr for `Number` types by holding a reference utill `ccall` returns (fixes #12)
- allow using any immutable bits types to be used in keys and values (fixes #13)
- added some test cases

tanmaykm committed:
improve MDBValue type handling

- prevent possible invalid ptr for `Number` types by holding a reference utill `ccall` returns (fixes #12)
- allow using any immutable bits types to be used in keys and values (fixes #13)

tanmaykm commented:
looks like this needs UnalignedVectors?

wildart commented:
You could try to reinterpret array with the number to `UInt8` upon saving, and reinterpret back to `T` when reading.

tanmaykm committed:
address pointer not properly aligned error

tanmaykm committed:
prevent overriding generic `open`

Method with signature `open(f::Function, path::String)` interferes with the regular file open function in `Base`.

This commit will introduce a new type `MDBPath` to specialize for LMDB when needed.

wildart commented on test/cur.jl:
```diff
@@ -34,7 +34,7 @@ module LMDB_CUR
 
     # Block style
-    open(dbname) do env # open environment
+    open(LMDBPath(dbname)) do env # open environment
```
It is quite ugly to wrap a string into an object just to pass it as an argument. Maybe, you should rename `open` function to `environment`. This will give clear description of what function does.

tanmaykm replied:
Yeah... I too felt `open` was kind of too generic. A new exported function may be better instead.

wildart commented on src/cur.jl:
```diff
@@ -65,7 +65,8 @@ end
 """
-function get{T}(cur::Cursor, key, ::Type{T}, op::CursorOps=FIRST)
+get{K,T}(cur::Cursor, key::K, ::Type{T}, op::CursorOps=FIRST) = get(cur, [key], T)
+function get{K<:Union{String,Array},T}(cur::Cursor, key::K, ::Type{T}, op::CursorOps=FIRST)
```
Why put restriction on a key type?

tanmaykm replied:
There is no restriction. Just that the some key types need to be wrapped in an `Array` to be passed as a pointer (in `MDBValue`) to `ccall`.

wildart replied:
As I wrote in comment to `src/common.jl`, database interface should be invariant to key/value type.

wildart commented on src/common.jl:
```diff
@@ -1,29 +1,31 @@
+MDBValue(_::Void) = MDBValue()
+MDBValue{T<:Union{String,Array}}(val::T) = MDBValue(sizeof(val), pointer(val))
+function MDBValue{T}(val::T)
+    if isbits(T)
```
I still want a bit type value to be used. What if we create a field in `MDBValue` of value type to keep copy of the value and anchor it against GC sweeps.
```julia
struct MDBValue{T}
      size::Csize_t   # size of the data item
      data::Ptr{Void} # address of the data item
      value::T
end

# this is a general case
function MDBValue(v::T) where T
    o = MDBValue(Csize_t(sizeof(v)), C_NULL, v)
    o.data = pointer(o.value)
    return o
end
# this on for bit types
function MDBValue(v::T) where {T <: Number}
    o = MDBValue(Csize_t(sizeof(v)), C_NULL, T[v])
    o.data = pointer(o.value)
    return o
end
```

tanmaykm replied:
Setting bits type data is still possible in the current model.

This should work too... I tried to avoid this as the struct is passed into C (but should be okay as the additional data is at the end). I can give it a shot.

When `T <: Number`, `value` is `Vector{T}` and not `T`. So maybe `value::Union{T,Vector{T}}` is what we need.

wildart replied:
I think that `MDBVaule` should only work with arrays, strings and immutable structs. The bits types can be wrapped in the `put!`/`get` call into the array before constructing `MDBValue` object. Thus there will be no difference between bit type and 1-element array of such type.

wildart reviewed (commented):
Handling of bit types in `MDBValue` class

tanmaykm commented:
Thanks @wildart. #14 looks good to me.
Please feel free to close this.

wildart commented:
@tanmaykm I like your test overhaul. That should be next PR.
